Remove many `CompatibilitySwitch<T>` properties

dougbu opened the pull request:
- #4866 2 of 3
- will leave one `IEnumerable<ICompatibilitySwitch>` implementation: `MvcOptions`
- will leave one `ConfigureCompatibilityOptions<MvcOptions>` subclass: `MvcOptionsConfigureCompatibilityOptions`
- a few options remain as regular properties:
  - `ApiBehaviorOptions.SuppressMapClientErrors` (default `false`)
  - `MvcOptions.EnableEndpointRouting` (default `true`)
  - `MvcOptions.MaxValidationDepth` (default `32`)
  - `MvcJsonOptions.AllowInputFormatterExceptionMessages` (default `true`)
- remains possible to disable many features
  - e.g. replace `IModelBinderProvider`s with ones that pass `allowValidatingTopLevelNodes: false`

dougbu committed:
Remove many `CompatibilitySwitch<T>` properties
- #7156 2 of 3
- will leave one `IEnumerable<ICompatibilitySwitch>` implementation: `MvcOptions`
- will leave one `ConfigureCompatibilityOptions<MvcOptions>` subclass: `MvcOptionsConfigureCompatibilityOptions`
- a few options remain as regular properties:
  - `ApiBehaviorOptions.SuppressMapClientErrors` (default `false`)
  - `MvcOptions.EnableEndpointRouting` (default `true`)
  - `MvcOptions.MaxValidationDepth` (default `32`)
  - `MvcJsonOptions.AllowInputFormatterExceptionMessages` (default `true`)
- remains possible to disable many features
  - e.g. replace `IModelBinderProvider`s with ones that pass `allowValidatingTopLevelNodes: false`

dougbu commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/InferParameterBindingInfoConvention.cs:
```diff
@@ -118,13 +116,7 @@ private bool ParameterExistsInAnyRoute(ActionModel action, string parameterName)
-            var metadata = _modelMetadataProvider
-                .GetMetadataForType(parameter.ParameterInfo.ParameterType);
-
-            if (AllowInferringBindingSourceForCollectionTypesAsFromQuery && metadata.IsCollectionType)
```
`AllowInferringBindingSourceForCollectionTypesAsFromQuery` was one of the few properties that was `false` with recent `CompatibilityVersion`s.

JamesNK commented:
> * MvcOptions.EnableEndpointRouting (default true)

I'm not super familiar with compat but this point is fine 👍

dougbu commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -136,38 +134,30 @@ private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFil
             var canCache = PolicyProvider == null;
 
-            if (_mvcOptions == null) 
+            if (!context.IsEffectivePolicy(this))
```
Not sure why this check was behind the `AllowCombiningAuthorizeFilters` option. That required some seemingly-unrelated test changes.

In any case, should this be above the `ComputePolicyAsync()` call or in `OnAuthorizationAsync(...)`, prior to calling `GetEffectivePolicyAsync(...)`. The extra `async` state machines for this sync call seem wasteful.

pranavkm replied:
Yeah, that seems wonky. There's an issue tracking cleaning this type up in 3.0: https://github.com/aspnet/AspNetCore/issues/4862

rynowak replied:
👍

rynowak replied:
Yeah I would imagine so.

dougbu replied:
@rynowak should I reorder things in this PR or just add a comment in #4862?

dougbu commented on src/Mvc/test/Microsoft.AspNetCore.Mvc.Core.Test/Authorization/AuthorizeFilterTest.cs:
```diff
@@ -31,13 +31,15 @@ public async Task DefaultConstructor_DeniesAnonymousUsers()
             var filterFactory = authorizeFilterFactory as IFilterFactory;
             var authorizeFilter = (AuthorizeFilter)filterFactory.CreateInstance(
                 authorizationContext.HttpContext.RequestServices);
+            authorizationContext.Filters.Add(authorizeFilter);
```
Now required because the filter checks whether it's the effective one.

dougbu commented on src/Mvc/test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs:
```diff
@@ -1718,11 +1719,12 @@ public async Task FromBody_JToken_ExcludedFromValidation()
                     request.ContentType = "application/json";
-                });
+                },
+                mvcOptions: options);
```
Need to use the same `MvcOptions` instance was an odd fallout of removing `AllowValidatingTopLevelNodes`. The `ParameterBinderExtensions` helpers create `ModelMetadata` with a provider that uses these options. The default `MvcOptions` that `GetTestContext(...)` returns does not configure any binder providers.

dougbu commented:
:up::date: just to rebase
@pranavkm & @rynowak please have a look. This PR is the only thing that's holding me up opening up one for #4866 part 3 of 3.

pranavkm commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs:
```diff
@@ -45,11 +29,8 @@ public void Configure(ApiBehaviorOptions options)
-        public override void PostConfigure(string name, ApiBehaviorOptions options)
+        public void PostConfigure(string name, ApiBehaviorOptions options)
         {
-            // Let compatibility switches do their thing.
```
In 3.0, `InvalidModelStateResponseFactory` would default to returning a `ProblemDetails` instance (i.e. `options.InvalidModelStateResponseFactory = ProblemDetailsFactory;`) and `SuppressMapClientErrors` would have no bearing on it. Any user configured factory would take precedence over the default

dougbu replied:
@pranavkm I don't disagree. But, are you asking for a change here?

pranavkm replied:
Basically, set `options.InvalidModelStateResponseFactory = ProblemDetailsFactory;` in `Configure` and remove the `PostConfigure` since it's no longer conditional.

dougbu replied:
The `InvalidModelStateResponseFactory` setting remains conditional on `SuppressMapClientErrors` and that's a regular option that may be set at any point. Better to leave reading it 'til `PostConfigure(...)`.

pranavkm commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs:
```diff
@@ -53,7 +52,7 @@ public class ParameterBinder
         /// <param name="modelBinderFactory">The <see cref="IModelBinderFactory"/>.</param>
         /// <param name="validator">The <see cref="IObjectModelValidator"/>.</param>
-        /// <param name="mvcOptions">The <see cref="MvcOptions"/> accessor.</param>
+        /// <param name="mvcOptions">The <see cref="MvcOptions"/> accessor. This parameter is currently unused.</param>
```
I'd prefer leaving this doc comment as is. It'd be easy to miss updating this if we do start using it.

dougbu replied:
It's not clear to me this parameter will be used in the future. But, I could move this to `<remarks>`. Either way, the comment is hard to miss.

pranavkm commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs:
```diff
@@ -39,10 +39,12 @@ internal class DefaultObjectValidator : ObjectModelValidator
+                validationState)
+            {
+                MaxValidationDepth = _mvcOptions.MaxValidationDepth,
+                AllowShortCircuitingValidationWhenNoValidatorsArePresent =
```
Might have missed when this was discussed in the issue, but would we keep this? I thought we would just default to doing this in 3.0. The flag was there because we just didn't feel confident enough about enabling the behavior super late in 2.2

dougbu replied:
The change here is just cleanup. `AllowShortCircuitingValidationWhenNoValidatorsArePresent` goes away in part 3 of 3.

pranavkm commented on src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs:
```diff
@@ -67,7 +67,7 @@ public async Task AuthorizationPoliciesCombine()
             request.Headers.Add("Cookie", cookie2);
 
-            // Act 2: Will succeed because, with AllowCombiningAuthorizeFilters true, [Authorize] allows either cookie.
+            // Act 2: Will succeed because {Authorize] allows either cookie.
```
`[Authorize]`

dougbu replied:
Is there a typo here I can't see?

pranavkm replied:
There's a brace `{` at the start instead of `[`, no?

dougbu replied:
Got it

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs:
```diff
@@ -197,7 +197,7 @@ public virtual bool CanCreateInstance(Type targetType)
         /// <param name="bindingContext">The <see cref="ModelBindingContext"/>.</param>
         /// <remarks>
         /// <para>
-        /// This method should be called only when <see cref="MvcOptions.AllowValidatingTopLevelNodes" /> is
```
Should we also remove the implementations that handle `AllowValidatingTopLevelNodes` and just mark it obsolete and say that it doesn't do anything?

dougbu replied:
@rynowak why make the additional breaking change?

Users can disable validation of top-level nodes in a piecemeal fashion:
- replace the `IModelBinderProvider`s
- replace the `IModelMetadataProvider` with something that doesn't extend `ModelMetadataProvider`, and / or
- replace the `IObjectModelValidator` with something that doesn't extend `ObjectModelValidator`

But, ignoring `AllowValidatingTopLevelNodes` in the model binders would mean top-level validation could only be partially disabled.

rynowak replied:
We don't intend to allow you to disable validation for parameters. If you don't like it, don't put validation attributes on parameters. We did a compat switch for this in case we broke functionality - which we did 😆 

I'm not suggesting we break lots of APIs. I'm suggesting that where the property/arg appears, we ignore it and document that it is ignored.

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json:
```diff
@@ -72,5 +72,35 @@
+  {
+    "TypeId": "public class Microsoft.AspNetCore.Mvc.MvcOptions : System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Mvc.Infrastructure.ICompatibilitySwitch>",
+    "MemberId": "public System.Boolean get_AllowBindingHeaderValuesToNonStringModelTypes()",
+    "Kind": "Removal"
```
We should consider leaving these implementations here, **or** coming up with a better design. I don't really relish us adding and removing this (and churning supressions) each release. 

Adding/removing properties is semantically meaningful, but this isn't

dougbu replied:
This is about removing properties (specifically their getters and setters). We decided in an offline email thread to
> Completely remove the properties we no longer want to support i.e. those enabling most of the legacy behaviours.

Are you suggesting reopening that conversation?

rynowak replied:
Maybe I'm confused. I thought you also un-implemented the interfaces.

rynowak replied:
This is what I'm talking about: https://github.com/aspnet/AspNetCore/pull/4628/files#diff-d9489227febd545e09ee512b868b9b8dL16

dougbu committed:
PR comments
- move doc comments about ignored parameters to `<remarks>` and (in one case) `<exception>`
- move `IsEffectivePolicy(...)` check earlier in `AuthorizeFilter`
- ignore `allowValidatingTopLevelNodes` parameters and `AllowValidatingTopLevelNodes` properties in model binders
  - but, used in `CollectionModelBinder<TElement>` so that `DictionaryModelBinder<TKey, TValue>` can delay the check

dougbu commented:
@rynowak a couple of questions to you are outstanding&hellip;

For `[Authorize]`:
> should I reorder things in this PR or just add a comment in #4862?

For the model binders: Should I ignore all `allowValidatingTopLevelNodes` parameters and `AllowValidatingTopLevelNodes` properties except the pair in `CollectionModelBinder`? That would be a single inconsistency instead of
> the only parameter and property pair that could be ignored are in ComplexTypeModelBinder and DictionaryModelBinder. The inconsistency would be a bit ugly.

dougbu commented:
@pranavkm I couldn't understand a couple of your comments&hellip;

For `ApiBehaviorOptionsSetup`:
> In 3.0, `InvalidModelStateResponseFactory` would default to returning a `ProblemDetails` instance (i.e. `options.InvalidModelStateResponseFactory = ProblemDetailsFactory;`) and `SuppressMapClientErrors` would have no bearing on it. Any user configured factory would take precedence over the default

For `GlobalAuthorizationFilterTest`:
> `[Authorize]`

dougbu committed:
Revert removal of `IEnumerable<ICompatibilitySwitch>` implementations

dougbu committed:
PR comments: Correct typo

dougbu commented:
:up::date: to address one of @pranavkm's suggestions and to rebase
